go/p2p/peermgmt/registry.go

```diff
@@ -0,0 +1,277 @@
+	if err != nil {
+		r.logger.Error("failed to watch registry for node list changes",
+			"err", err,
+		)
+		return
+	}
+	defer nlSub.Close()
+
+	// Listen to nodes on node events.
+	nodeCh, nSub, err := r.consensus.Registry().WatchNodes(ctx)
+	if err != nil {
+		r.logger.Error("failed to watch registry for node changes",
+			"err", err,
+		)
+		return
+	}
+	defer nSub.Close()
+
+	for {
+		select {
+		case nodes := <-nodeListCh:
+			r.handleNodes(nodes.Nodes, true)
+
+		case nodeEv := <-nodeCh:
+			if nodeEv.IsRegistration {
+				r.handleNodes([]*node.Node{nodeEv.Node}, false)
+			}
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// handleNodes updates protocols and topics supported by the given nodes and resets them if needed.
+func (r *peerRegistry) handleNodes(nodes []*node.Node, reset bool) {
+	defer r.initOnce.Do(func() {
+		close(r.initCh)
+	})
+
+	type peerData struct {
+		info      *peer.AddrInfo
+		protocols []core.ProtocolID
+		topics    []string
+	}
+
+	peers := make(map[core.PeerID]*peerData)
+	for _, n := range nodes {
+		info, err := p2pInfoToAddrInfo(&n.P2P)
+		if err != nil {
+			r.logger.Error("failed to convert node to node info",
+				"err", err,
+				"node_id", n.ID,
+			)
+			continue
+		}
+
+		protocols, topics := r.inspectNode(n)
+
+		peers[info.ID] = &peerData{info, protocols, topics}
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// Remove previous state.
+	if reset {
+		r.peers = make(map[core.PeerID]*peer.AddrInfo)
+		r.protocolPeers = make(map[core.ProtocolID]map[core.PeerID]*peer.AddrInfo)
+		r.topicPeers = make(map[string]map[core.PeerID]*peer.AddrInfo)
+	}
+
+	// Add/update new peers.
+	for p, data := range peers {
+		// Remove old protocols/topics.
+		for _, peers := range r.protocolPeers {
+			delete(peers, p)
+		}
+		for _, peers := range r.topicPeers {
+			delete(peers, p)
+		}
+
+		// Add new ones.
+		for _, protocol := range data.protocols {
+			peers, ok := r.protocolPeers[protocol]
+			if !ok {
+				peers = make(map[core.PeerID]*peer.AddrInfo)
+				r.protocolPeers[protocol] = peers
+			}
+			peers[p] = data.info
+		}
+		for _, topic := range data.topics {
+			peers, ok := r.topicPeers[topic]
+			if !ok {
+				peers = make(map[core.PeerID]*peer.AddrInfo)
+				r.topicPeers[topic] = peers
+			}
+			peers[p] = data.info
+		}
+
+		// Update the address, as it might have changed.
+		r.peers[p] = data.info
+	}
+}
+
+func (r *peerRegistry) inspectNode(n *node.Node) ([]core.ProtocolID, []string) {
+	// Protocol versions cannot be found in the registry so we assume all peers speak the latest
+	// protocols. The peer manager should handle this problem and block or negatively grade peers
+	// that are not up-to-date.
+	protocols := make([]core.ProtocolID, 0)
+	topics := make([]string, 0)
+
+	if n.HasRoles(node.RoleKeyManager) {
+		for _, rt := range n.Runtimes {
+			pid := rpc.NewRuntimeProtocolID(rt.ID, kmp2p.KeyManagerProtocolID, kmp2p.KeyManagerProtocolVersion)
```

Comment:
Seems a bit backwards that the general P2P stack (in `go/p2p`) needs to import all of the higher-level P2P protocols for this discovery purpose? Would it be better to have some sort of global registry that would be populated during `init()` inside these higher-level protocols (e.g. the higher-level protocol would import this package and invoke some function to add a handler).

The handler would get passed a node and would return the protocols/topics (or, maybe better, would add protocols/topics to some existing map that would be passed).

Reply:
Agree totally. Fixed. I prefer not to pass maps as the handler could manipulate it or produce some kind of data races.